### PR TITLE
consolidate segment setup

### PR DESCRIPTION
**Pull request by brad-decker**
in #9509 a good step was made to move segment setup up from one of the controllers in the background process. This goes a step further by deduplicating a big chunk of setup between the frontend and background.

<details>
<summary>Bug fix, categories aren't attached correctly for background events</summary>

</details>

### Improvements
1. adds environment_type to background events
2. adds chain_id to all events
3. updates network to either be the `rpcUrl`, if `provider.type` is `rpc`, or just `provider.type` otherwise.

### CODE REVIEW

File: shared/modules/metametrics.js
Comment (rekmarks) [inline shared/modules/metametrics.js:165]: Do we need the `omit` call here? The keys omitted from `properties` will be overwritten anyway since they are explicitly specified below.

File: shared/modules/metametrics.js
Comment (rekmarks) [reply shared/modules/metametrics.js:49]: I think it's either missing or set to the string `'true'`.

File: ui/app/selectors/selectors.js
Comment (rekmarks) [inline ui/app/selectors/selectors.js:26]: We should use this in the `getAddressBook` selector in the same file.

File: ui/app/selectors/selectors.js
Comment (rekmarks) [inline ui/app/selectors/selectors.js:18]: Personal preference: I don't like nested destructuring assignment; I find it hard to read. So, I will suggest the change, but not that it's optional. In this case, it _would_ make it more consistent with `getCurrentChainId` below:
```suggestion
  const { provider } = state.metamask
```

File: shared/modules/metametrics.js
Comment (rekmarks) [inline shared/modules/metametrics.js:146]: Looking at the [`ConfirmTransactionBase` component](https://github.com/MetaMask/metamask-extension/blob/develop/ui/app/pages/confirm-transaction-base/confirm-transaction-base.component.js#L475-L488), aren't we racing to increment the `metaMetricsSendCount`?

This may be a preexisting problem, but we should increment and update `metaMetricsSendCount` in a single place, ideally in some MetaMetrics handler function.

It's OK if we do that in a separate PR, but we should absolutely do it.

**Review by rekmarks — CHANGES_REQUESTED** [review]
I'm a big fan of this change!

My comments are mostly nits and questions, but there's on thing re: the `metaMetricsSendCount` that we need to address in this or (I think I'd prefer) a later PR.

File: app/scripts/lib/rpc-method-middleware/handlers/log-web3-usage.js
Comment (brad-decker) [reply app/scripts/lib/rpc-method-middleware/handlers/log-web3-usage.js:57]: In the previous iteration, `referrerUrl` was being mapped to context.referrer.url:

image

File: app/scripts/metamask-controller.js
Comment (brad-decker) [reply app/scripts/metamask-controller.js:125]: lol noooo. I went through a few iterations of api design before landing on this and at one point these two things were in separate functions 👀 lol. I, too, prefer order over chaos and I will fix this to avoid complicating the git history.

File: shared/modules/metametrics.js
Comment (brad-decker) [reply shared/modules/metametrics.js:24]: Thanks!

File: shared/modules/metametrics.js
Comment (brad-decker) [reply shared/modules/metametrics.js:105]: Hmm, no it's not. 😬 -- I'll switch it to void.

File: app/scripts/metamask-controller.js
Comment (rekmarks) [reply app/scripts/metamask-controller.js:125]: 😂

File: shared/modules/metametrics.js
Comment (brad-decker) [reply shared/modules/metametrics.js:165]: the keys below come from a different source. If someone tries to use these properties, which have special meaning in segment, I want to make them think about the usage. Ideally, when typescript comes around they'll get an immediate error when passing these keys to properties object, but for now, their payload will *not* include the values. When they investigate they will see that these three are special cases and should be elevated outside of the properties object. They might also realize that 'value' isn't an appropriate place to put the data they are tracking.

File: shared/modules/metametrics.js
Comment (brad-decker) [reply shared/modules/metametrics.js:146]: I'll address this in a separate PR

**Comment by brad-decker** [comment]
@rekmarks I addressed your feedback in the latest commit

File: shared/modules/metametrics.js
Comment (rekmarks) [reply shared/modules/metametrics.js:164]: > When they investigate they will see that these three are special cases

I think we could achieve the same effect with a comment!

File: app/scripts/metamask-controller.js
Comment (rekmarks) [inline app/scripts/metamask-controller.js:119]: Oh, one more: I think we can remove the comment, which is useless, and then we should probably rename `MetaMaskController.trackSegmentEvent` to `trackMetaMetricsEvent`.

File: ui/app/selectors/selectors.js
Comment (rekmarks) [inline ui/app/selectors/selectors.js:17]: Looking at the `getNetworkIdentifier` selector above, the naming here is a bit confusing.

Can we rename this to `getMetricsNetworkIdentifier` to better indicate its intended use?

File: shared/modules/metametrics.js
Comment (brad-decker) [reply shared/modules/metametrics.js:168]: I added a comment above the omit. I feel strongly that we should avoid treating these as normal properties, but adding additional comments about why they are being omitted makes sense to me.

**Comment by brad-decker** [comment]
made those requested changes @rekmarks! thanks for the feedback.

**Review by rekmarks — DISMISSED** [review]
LGTM!

**Comment by brad-decker** [comment]
Waiting for a review from @Gudahtt before I merge as he expressed interest in doing so; I want to get additional buy-in on the approach as it's the first shared module. Thanks for your reviews, @rekmarks. I've already incorporated these changes into my WIP branches off of this one to great effect.

File: shared/modules/metametrics.js
Comment (Gudahtt) [inline shared/modules/metametrics.js:191]: This.... is unfortunate. We have situations where we need to send an event immediately before closing the notification window, which ends the process :grimacing:

Hmm. I guess we can cross that bridge when we come to it, perhaps by sending those from the background.

Comment (Gudahtt) [reply shared/modules/metametrics.js:191]: That said, I'm not sure this is particularly useful? I don't think the timing is critical in any situation except those where the process is about to end, and in that situation it sounds like this still doesn't provide any guarantee.

File: shared/modules/metametrics.js
Comment (Gudahtt) [inline shared/modules/metametrics.js:153]: I found it a bit confusing trying to keep track of which piece of context was passed in where. I'm not sure offhand how this can be improved, but allowing _either_ of these two values to contain _any_ part of the context object doesn't seem ideal here.

Right now if `app` or `userAgent` are provided, they'd be overwritten for example. Perhaps more importantly, the type for `providedContext` and `additionalContext` doesn't communicate that those properties are not expected to be passed in. And the same problem exists with `providedContext` vs `additionalContext`.

Maybe they can be split into two distinct types, and combined here? e.g. `MetaMetricsPageContext` and `MetaMetricsEventContext` - something like that.

File: shared/modules/metametrics.js
Comment (brad-decker) [reply shared/modules/metametrics.js:103]: The default was set to true in the background's version of the segment controller. If we want to land on this being false and have the background manually override, then I'm okay with removing it!

File: shared/modules/metametrics.js
Comment (brad-decker) [reply shared/modules/metametrics.js:191]: that all makes sense to me. One other factor that made me go this route was that we already had places that were using `.then` on the analytics call.

Does the beforeunload event handler work in the background as well? I just saw node-analytics has a manual `.flush` that I can call that'll resolve the segment issue.

Comment (Gudahtt) [reply shared/modules/metametrics.js:191]: Interesting! That sounds promising.

I'm not even sure if `beforeunload` works in the popup. I haven't verified it anyway. I'd also be curious to know whether it works in the background.

Oh and this whole thread was non-blocking for this PR - we can improve this later if that's easier.

Comment (brad-decker) [reply shared/modules/metametrics.js:191]: `metametrics-opt-in.component.js` is the example I was thinking of. I wanted to confirm my memory wasn't playing tricks on me.

```js
const promise = participateInMetaMetrics === true
  ? metricsEvent({
    eventOpts: {
      category: 'Onboarding',
      action: 'Metrics Option',
      name: 'Metrics Opt Out',
    },
    isOptIn: true,
  })
  : Promise.resolve()

promise
  .then(() => {
    history.push(nextRoute)
  })
```

File: shared/modules/metametrics.js
Comment (Gudahtt) [reply shared/modules/metametrics.js:103]: Hmm. Was, but is no longer?

I think a default of `false` makes sense. That is the more common case :shrug:

Comment (brad-decker) [reply shared/modules/metametrics.js:103]: works for me!

File: shared/modules/metametrics.js
Comment (Gudahtt) [reply shared/modules/metametrics.js:191]: Interesting - actually now that you mention it, I can see the case for blocking on that in some cases. It is particularly important that we get that particular event.

Comment (brad-decker) [reply shared/modules/metametrics.js]: I did some testing and 'beforeunload' does not happen for notification windows. 'unload', does though but we get no guarantees that `segment.flush` will finish and do its thing. I have made a call to make `flushAt: 1` for all environments instead of just development. It doesn't seem like we'd be capturing all of the events in notifications.

File: shared/modules/metametrics.js
Comment (brad-decker) [reply shared/modules/metametrics.js]: I took a stab at making this better. essentially:
the context key provided to the 'getMetaMetricsTrackEvent' can only include page or referrer. The same is true for context provided to `eventContext` (previously `additionalContext`). I also pick those keys out in case someone doesn't follow the types 😱. Locale is now a top level key of `MetaMetricsRequiredState` instead of being a subkey of context. 

let me know if this is clearer.

File: shared/modules/metametrics.js
Comment (brad-decker) [reply shared/modules/metametrics.js]: This is represented in the latest work

File: shared/modules/metametrics.js
Comment (brad-decker) [reply shared/modules/metametrics.js]: the default was supposed to be true for background, but i had set it to false. This is fixed now and all calls in the background currently specify excludeMetaMetricsId.

Comment (Gudahtt) [reply shared/modules/metametrics.js]: Ah I see - that makes sense. Thanks, this is much better

File: shared/modules/metametrics.js
Comment (Gudahtt) [reply shared/modules/metametrics.js]: Good to know! This seems like a great solution.

I can understand why they'd want to batch events together by default but our usage is far from typical. We can turn the batching back on in the background specifically (and maybe in the fullscreen UI?) if this ever becomes a concern.

File: shared/modules/metametrics.js
Comment (Gudahtt) [reply shared/modules/metametrics.js]: Much clearer!

**Review by Gudahtt — APPROVED** [review]
LGTM!

**Pull request merged**
